graph-http/src/iotools.rs

```diff
@@ -1,62 +1,86 @@
-        }
-        Ok(())
-    }
-
-    pub async fn create_dir_async<P: AsRef<Path>>(directory: P) -> GraphResult<()> {
-        if !directory.as_ref().exists() {
-            tokio::fs::create_dir_all(directory).await?;
-        }
-        Ok(())
+pub fn create_dir<P: AsRef<Path>>(directory: P) -> Result<(), std::io::Error> {
+    if !directory.as_ref().exists() {
+        fs::create_dir_all(&directory)?;
     }
+    Ok(())
+}
 
-    pub fn copy(path: PathBuf, mut response: reqwest::blocking::Response) -> GraphResult<PathBuf> {
-        let (sender, receiver) = mpsc::channel();
-        let handle = thread::spawn(move || {
-            let mut file_writer = OpenOptions::new()
-                .create(true)
-                .write(true)
-                .read(true)
-                .open(&path)
-                .expect("Error creating file");
-            copy(&mut response, &mut file_writer).expect("Error copying file contents");
-            sender.send(Some(path)).unwrap();
-        });
-
-        handle.join().expect("Thread could not be joined");
-        match receiver.recv() {
-            Ok(t) => {
-                Ok(t.ok_or_else(|| GraphFailure::not_found("Unknown error downloading file"))?)
-            },
-            Err(e) => Err(GraphFailure::from(e)),
-        }
+pub async fn create_dir_async<P: AsRef<Path>>(directory: P) -> Result<(), std::io::Error> {
+    if !directory.as_ref().exists() {
+        tokio::fs::create_dir_all(directory).await?;
     }
+    Ok(())
+}
 
-    pub async fn copy_async(path: PathBuf, response: reqwest::Response) -> GraphResult<PathBuf> {
-        let mut file = tokio::fs::OpenOptions::new()
+pub fn copy(
+    path: PathBuf,
+    mut response: reqwest::blocking::Response,
+) -> Result<PathBuf, ThreadedIoError> {
+    let (sender, receiver) = mpsc::channel();
+    let handle = thread::spawn::<_, Result<(), ThreadedIoError>>(move || {
+        let mut file_writer = fs::OpenOptions::new()
             .create(true)
             .write(true)
             .read(true)
-            .open(&path)
-            .await?;
-        let mut stream = response.bytes_stream();
-        while let Some(item) = stream.next().await {
-            file.write_all(&item?).await?;
-        }
-        Ok(path)
+            .open(&path)?;
+        std::io::copy(&mut response, &mut file_writer)?;
+        sender.send(Some(path))?;
+        Ok(())
+    });
+
+    handle.join().map_err(ThreadedIoError::Join)??;
+    receiver.recv()?.ok_or(ThreadedIoError::NoPath)
+}
+
+pub async fn copy_async(
+    path: PathBuf,
+    response: reqwest::Response,
+) -> Result<PathBuf, AsyncIoError> {
+    let mut file = tokio::fs::OpenOptions::new()
+        .create(true)
+        .write(true)
+        .read(true)
+        .open(&path)
+        .await?;
+    let mut stream = response.bytes_stream();
+    while let Some(item) = stream.next().await {
+        file.write_all(&item?).await?;
     }
+    Ok(path)
+}
+
+#[derive(Debug, thiserror::Error)]
```

Comment:
I have some concerns about these errors that you have added. My main concern is that I am unsure what adding these will benefit the project or those that use it. Do they provide more info on the error? I can possibly see that having these may indicate what type of error it is but is that really any different than what was already in place? Just looking for some context here.

Reply:
I think that's a fair criticism. It may be something of a philosophical difference/preference, and since you're the owner/maintainer of this tool, I'll go with whatever you prefer. Sorry in advance for the wall of text, but I've been thinking a lot about this recently in other contexts and wanted to get some thoughts down.

---

## Typestate

One of Rust's great strengths, imo, is the "correctness" aspect of the type system. You can construct your code such that invalid states are actually impossible. For example, `iotools::copy()` before returned a `GraphFailure` error, but it is impossible for it to generate a `serde_yaml::Error` or `handlebars::RenderError`. While it's cleaner in the library code, a consumer of the library would still need to match on all of those cases to either recover or panic.

Granted, in the case of `iotools`, an api consumer isn't going to use it directly, but more likely though one of the download clients. However, the story there is the same and we need to wrap the output of `iotools::copy` anyway to return a single error type.

Additionally, handling all the possible errors in `iotools::copy` without `panic`ing on threading errors with `GraphFailure` meant adding two new variants to `GraphFailure` which would never be relevant to any other error producer. Especially in a library, we probably want to let the api consumer decide when to `panic`/`expect`/`unwrap`. Searching through the library, there are a fair number of other panics that we could probably get rid of with a similar approach.

## Library Errors

A good example of the different approaches is in the difference between the target audiences of error crates, like [`anyhow`](https://github.com/dtolnay/anyhow) and [`thiserror`](https://github.com/dtolnay/thiserror) (both written by the same author, giving [this](https://github.com/dtolnay/thiserror#comparison-to-anyhow) comparison). `anyhow` is targeted at application errors, receiving just about any error type and presenting it to the user (and probably exiting without recovery). `thiserror` is more focused at nesting/combining error types for libraries, so that the consumer can decide what to do with them.

It feels like `GraphFailure` is acting a little more like an application error than a library error, and this is probably a first step at taking it the other direction. I don't think all of those changes will be part of this PR, but probably something to open for a future issue.

Reply:
Im open to trying new things, and I like the way you presented it so lets keep it.